crates/orderbook/src/generator.rs

```diff
@@ -0,0 +1,164 @@
+use futures::{channel::mpsc, stream::FusedStream, SinkExt, Stream};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Creates a stream from an async generator function.
+///
+/// This is similar to [`unfold`](https://docs.rs/futures/0.3.21/futures/stream/fn.unfold.html) in
+/// the sense that it bridges Future to Stream. In addition, it can be more ergonomic because it
+/// allows the compiler to automatically create the state machine that you would otherwise have to
+/// write by hand in the same way that async/await syntax makes it easier to create a Future.
+///
+/// The generator takes an argument `Sender` who's `send` function is used to yield values to the
+/// Stream.
+///
+/// The generator must ensure that the sender does not outlive the generator. This is usually the
+/// case but could be violated by leaking the sender or moving it into a global task. If you want to
+/// use a task then you do not need this function because you can communicate over a channel
+/// directly.
+pub fn async_generator_to_stream<T, Fut>(
+    generator: impl FnOnce(Sender<T>) -> Fut,
+) -> impl Stream<Item = T> + FusedStream
+where
+    Fut: Future<Output = ()>,
+{
+    let (sender, receiver) = mpsc::channel(0);
+    let generator = generator(Sender(sender));
+    StreamAndFuture::new(receiver, Some(generator))
+}
+
+pub struct Sender<T>(mpsc::Sender<T>);
+
+impl<T> Sender<T> {
+    /// Yields a value from the generator into the stream.
+    pub async fn send(&mut self, t: T) {
+        // We assume that the receiver has not been dropped so sending cannot fail.
```

Comment:
> We **assume** that the receiver has not been dropped...

Don't we know that for sure?
The `StreamAndFuture` owns the receiver and the sender (at least by owning the generator) and doesn't expose access to them.
Also the `Sender` doesn't implement any trait which would allow it to be copied and moved out of the generator to outlive the `StreamAndFuture`, no?

Reply:
This is explained a bit in the documentation for `async_generator_to_stream`. One way of this happening is if you spawn a background task and move the sender into that task. Then StreamAndFuture can be dropped which drops the receiver but you still access to the sender.
`Sender` can still be moved even when it isn't Clone or Copy and I didn't find a reasonable way of preventing moves of it in the generator closure.

Reply:
Ahh, yes. Forgot about that.
Could that be prevented by requiring the generator to return its `Sender` on completion?
This is just a thought experiment because the worse ergonomics would probably not be worth it to prevent this edge case.
```rust
pub fn async_generator_to_stream<T, Fut>(
    generator: impl FnOnce(Sender<T>) -> Fut,
) -> impl Stream<Item = T> + FusedStream
where
    Fut: Future<Output = Sender<T>>,
```

Reply:
That's one of the ideas I had. It doesn't work because you can do something like
```rust
async fn generator(sender: Sender) -> Sender {
    move_sender_to_task(sender);
    futures::pending::<Sender>().await
}
```

Then StreamAndFuture can get dropped and you're back to having a sender with no receiver. This works because you can write functions like
```rust
fn foo() -> Sender<usize> {
    loop {}
}
```
that "return" any output type even those that have no way to be constructed because the compiler understands that the function never actually returns anything.

Reply:
Another idea is using `&mut Sender` but then lifetime become awkward and nothing can be Unpin because the pointed to Sender isn't allowed to move.

Reply:
Oh boy! Very interesting problem.

Another idea: AFAIK the type of every closure is unique. Maybe this fact plus the requirement that the generator returns what we put in could avoid the issue of people "faking" the return type (i.e. `futures::pending::<Sender>().await`).

One could imagine to pass a closure which does the sending instead of a type `Sender`:
```rust
pub fn async_generator_to_stream<T, Fut>(
    generator: impl FnOnce(SENDER) -> Fut,
) -> impl Stream<Item = SENDER> + FusedStream
where
    Fut: Future<Output = SENDER>,
    SENDER: Fn(T) -> ()
```

Reply:
While you cannot write out the type you still have access to it. If I make my generator a real function instead of a closure  it would look like
```rust
fn generator<T>(sender: T) where T: Fn(T) -> () {}
```
Inside of it I can still call the type T and call pending. Even if I was forced to write `sender: impl ...` I believe `pending` would still be able to infer the correct type.

Reply:
Of course. 🤦‍♂️ 
I had the order of "usability" of the function types backwards. I thought `FnOnce` would always require a closure (which can't use generic argument or return types AFAIK), but since all other function types are more generic they could be used as well.
In that case I have nothing to add here and might as well approve the PR. 😅

Reply:
You can also expose an additional `try_send` on `Sender` that bubbles up the error and have the generator handle it? This would allow moving the future to a background task and still handle the stream being dropped, while still keeping the `send` for ergonomics with simple generators?

> Another idea is using `&mut Sender` but then lifetime become awkward and nothing can be Unpin because the pointed to Sender isn't allowed to move.

Can you pass in a `Pin<&mut Sender>` and handle pinning in the caller?